networking/eth2/src/main/java/tech/pegasys/artemis/networking/eth2/gossip/topics/validation/AttestationValidator.java

```diff
@@ -0,0 +1,212 @@
+
+  private ValidationResult singleOrAggregateAttestationChecks(final Attestation attestation) {
+    // attestation.data.slot is within the last ATTESTATION_PROPAGATION_SLOT_RANGE slots (within a
+    // MAXIMUM_GOSSIP_CLOCK_DISPARITY allowance) -- i.e. attestation.data.slot +
+    // ATTESTATION_PROPAGATION_SLOT_RANGE >= current_slot >= attestation.data.slot (a client MAY
+    // queue
+    // future attestations for processing at the appropriate slot).
+    final UnsignedLong currentTimeMillis = secondsToMillis(recentChainData.getStore().getTime());
+    if (isAfterPropagationSlotRange(currentTimeMillis, attestation)) {
+      return INVALID;
+    }
+    if (isBeforeMinimumBroadcastTime(attestation, currentTimeMillis) < 0) {
+      return SAVED_FOR_FUTURE;
+    }
+
+    // The block being voted for (attestation.data.beacon_block_root) passes validation.
+    // It must pass validation to be in the store.
+    // If it's not in the store, it may not have been processed yet so save for future.
+    final Optional<BeaconState> maybeState =
+        recentChainData.getBlockState(attestation.getData().getBeacon_block_root());
+    if (maybeState.isEmpty()) {
+      return SAVED_FOR_FUTURE;
+    }
+
+    final BeaconState state = maybeState.get();
+
+    // The signature of attestation is valid.
+    final IndexedAttestation indexedAttestation = get_indexed_attestation(state, attestation);
+    if (!is_valid_indexed_attestation(state, indexedAttestation)) {
+      return INVALID;
+    }
+    return VALID;
+  }
+
+  private ValidatorAndSlot getValidatorAndSlot(final Attestation attestation) {
+    return new ValidatorAndSlot(
+        attestation.getData().getSlot(),
+        attestation.getData().getIndex(),
+        attestation.getAggregation_bits().streamAllSetBits().findFirst().orElseThrow());
+  }
+
+  private int isBeforeMinimumBroadcastTime(
+      final Attestation attestation, final UnsignedLong currentTimeMillis) {
+    final UnsignedLong minimumBroadcastTimeMillis =
+        minimumBroadcastTimeMillis(attestation.getData().getSlot());
+    return currentTimeMillis.compareTo(minimumBroadcastTimeMillis);
+  }
+
+  private boolean isAfterPropagationSlotRange(
+      final UnsignedLong currentTimeMillis, final Attestation aggregate) {
+    final UnsignedLong aggregateSlot = aggregate.getData().getSlot();
+    return maximumBroadcastTimeMillis(aggregateSlot).compareTo(currentTimeMillis) < 0;
+  }
+
+  private UnsignedLong secondsToMillis(final UnsignedLong seconds) {
+    return seconds.times(MILLIS_PER_SECOND);
+  }
+
+  private UnsignedLong minimumBroadcastTimeMillis(final UnsignedLong aggregateSlot) {
+    final UnsignedLong lastAllowedTime =
+        recentChainData
+            .getGenesisTime()
+            .plus(aggregateSlot.times(UnsignedLong.valueOf(SECONDS_PER_SLOT)));
+    final UnsignedLong lastAllowedTimeMillis = secondsToMillis(lastAllowedTime);
+    return lastAllowedTimeMillis.compareTo(MAXIMUM_GOSSIP_CLOCK_DISPARITY) >= 0
```

Comment:
The first check seems a bit redundant and makes the code hard to read since Unix time starts back in the 1970s

Reply:
Tests often default to starting at time 0 or small numbers because they're easier to read.